core/src/test/java/io/atomix/cluster/BalancingClusterManagerTest.java

```diff
@@ -0,0 +1,299 @@
+    List<AtomixReplica> active = createReplicas(3, clusterManager);
+    Cluster cluster = active.get(0).server().server().cluster();
+    cluster.members().stream().forEach(m -> {
+      m.onTypeChange(type -> {
+        if (type == Member.Type.PASSIVE) {
+          resume();
+        }
+      });
+    });
+    cluster.onJoin(m -> {
+      m.onTypeChange(type -> {
+        if (type == Member.Type.PASSIVE) {
+          resume();
+        }
+      });
+    });
+
+    // adding an ACTIVE member results in demotion of any ACTIVE member to PASSIVE
+    AtomixReplica joiner = createReplica(nextAddress(), AtomixReplica.Type.ACTIVE, clusterManager);
+    joiner.join(this.members).thenRun(this::resume);
+    await(10000, 2);
+  }
+
+  /**
+   * Tests demoting an active member to passive when active members are more than required and quorum hint is not {@link Quorum#ALL}
+   */
+  public void testDemotingActiveMemberToReserve() throws Throwable {
+    ClusterManager clusterManager = BalancingClusterManager.builder().withQuorumHint(3).withBackupCount(1).build();
+    createReplicas(3, clusterManager);
+
+    // add 2 passive member to the cluster to fulfill the required number of passive members
+    List<AtomixReplica> passives = join(clusterManager, AtomixReplica.Type.PASSIVE, AtomixReplica.Type.PASSIVE);
+    Cluster cluster = passives.get(0).server().server().cluster();
+    cluster.members().stream().filter(m -> m.type() == Member.Type.ACTIVE).forEach(m -> {
+      m.onTypeChange(type -> {
+        if (type == Member.Type.RESERVE) {
+          resume();
+        }
+      });
+    });
+    cluster.onJoin(m -> {
+      m.onTypeChange(type -> {
+        if (type == Member.Type.RESERVE) {
+          resume();
+        }
+      });
+    });
+
+    // now adding a PASSIVE member results in demotion of any ACTIVE member to RESERVE
+    AtomixReplica joiner = createReplica(nextAddress(), AtomixReplica.Type.ACTIVE, clusterManager);
+    joiner.join(this.members).thenRun(this::resume);
+    await(10000, 2);
+  }
+
+  /**
+   * Tests promoting an reserve member to passive when number of passive servers are less than required and quorum hint is not {@link Quorum#ALL}
+   */
+  public void testPromotingReserveMemberToPassive() throws Throwable {
+    ClusterManager clusterManager = BalancingClusterManager.builder().withQuorumHint(3).withBackupCount(1).build();
+    List<AtomixReplica> actives = createReplicas(3, clusterManager);
+    join(clusterManager, AtomixReplica.Type.PASSIVE);
+    Cluster cluster = actives.get(0).server().server().cluster();
+    cluster.onJoin(m -> {
+      m.onTypeChange(type -> {
+        if (type == Member.Type.PASSIVE) {
+          resume();
+        }
+      });
+    });
+
+    // adding a RESERVE member will result in promotion of this member to PASSIVE
+    AtomixReplica joiner = createReplica(nextAddress(), AtomixReplica.Type.RESERVE, clusterManager);
+    joiner.join(this.members).thenRun(this::resume);
+    await(10000, 2);
```

Comment:
Hard to see because I'm reviewing by mobile but where is await defined? Can we use a count down latch or join on the cf here instead?

Reply:
It's inherited from `ConcurrentTestCase`. It blocks the main test thread until some other thread calls `this::resume`.

Reply:
Okay fine as is then.